stonesoup/functions/interpolate.py

```diff
@@ -0,0 +1,116 @@
+import copy
+import datetime
+import warnings
+from typing import Union, List, Iterable
+
+import numpy as np
+
+from ..types.array import StateVectors
+from ..types.state import StateMutableSequence, State
+
+
+def time_range(start_time: datetime.datetime, end_time: datetime.datetime,
+               timestep: datetime.timedelta = datetime.timedelta(seconds=1)) \
+        -> Iterable[datetime.datetime]:
+    """
+    Produces a range of datetime object between ``start_time`` (inclusive) and ``end_time``
+    (inclusive)
+
+    Parameters
+    ----------
+    start_time: datetime.datetime   time range start (inclusive)
+    end_time: datetime.datetime     time range end (inclusive)
+    timestep: datetime.timedelta    default value is 1 second
+
+    Returns
+    -------
+    Generator[datetime.datetime]
+
+    """
+    duration = end_time - start_time
+    n_time_steps = duration / timestep
+    for x in range(int(n_time_steps) + 1):
+        yield start_time + x * timestep
+
+
+def interpolate_state_mutable_sequence(sms: StateMutableSequence,
+                                       times: Union[datetime.datetime, List[datetime.datetime]],
+                                       ) -> Union[StateMutableSequence, State]:
+    """
+    This function performs linear interpolation on a :class:`~.StateMutableSequence`. The function
+    has two slightly different forms:
+
+    If an individual :class:`~datetime.datetime` is inputted for the variable ``times`` then a
+    :class:`~.State` is returned corresponding to ``times``.
+
+    If a list of :class:`~datetime.datetime` is inputted for the variable ``times`` then a
+    :class:`~.StateMutableSequence` is returned with the states in the sequence corresponding to
+    ``times``.
+
+    Note
+    ----
+    This function does **not** extrapolate. Times outside the range of the time range of ``sms``
+    are discarded and warning is given. If all ``times`` values are outside the time range of
+    ``sms`` then an ``IndexError`` is raised.
+
+    Unique states for each time are required for interpolation. If there are multiple states with
+    the same time in ``sms`` the later state in the sequence is used.
+    """
+
+    # If single time is used, insert time into list and run again.
+    # A StateMutableSequence is produced by the inner function call.
+    # The single state is taken from that StateMutableSequence
+    if isinstance(times, datetime.datetime):
+        new_sms = interpolate_state_mutable_sequence(sms, [times])
+        return new_sms.state
+
+    # Track metadata removed and no interpolation can be performed on the metadata
+    new_sms = copy.copy(sms)
+    if hasattr(new_sms, "metadatas"):
+        new_sms.metadatas = list()
```

Comment:
Doc string should mention that `metadata` will be removed. Would be good to resolve this in future.